### ISSUE TITLE
Detached thread with server.listen() and execute server.stop() just after detach, fail to stop server

### ISSUE THREAD

**Issue by luismigb** [yhirose/cpp-httplib#1371]
I have a piece of code where I create a server and then a detached thread where the server.listen() runs indefinitely.

I have found that, if for any reason, server.stop() is called just after detaching thread, server.listen() does not set is_running_ variable in time, so when server.stop() is executed, the server is never stopped.

This is just a synchronization problem at high level usage, but maybe it can be managed in the library (some kind of handler to notify that the server is just running).

**Comment by yhirose** [comment]
Could you show me the smallest possible code to reproduce it? I may not able to fix it, but at least I would like to know the situation more fully. Thanks!

**Comment by luismigb** [comment]
Let see if you can replicate with this:

`class Test
{
public:
	void Start(void)
	{
		std::thread thread1(std::bind(&Test::Thread1, this));
		thread1.detach();

		std::thread thread2(std::bind(&Test::Thread2, this));
		thread2.detach();

		std::cout << "started" << std::endl;
	}
	void Thread1(void)
	{
		server1.Post();
		server1.listen();

		std::cout << "finished 1" << std::endl;
	}
	void Thread2(void)
	{
		server2.Post();
		server2.listen();

		std::cout << "finished 2" << std::endl;
	}

	Server server1;
	Server server2;
};

int main(void)
{
	Test test;

	test.Start();

	server2.stop();

	pause();
}`

**Comment by PixlRainbow** [comment]
This is just a race condition that occurs when you try to stop the server before the server actually starts.
By placing the call to "listen" in a detached thread, it becomes separately scheduled.
Even if your code semantically places a call to "stop" after a call to "listen", it doesn't mean that the call to "listen" will be executed immediately; the operating system may choose to schedule the call for later execution.

There are four options
1. add a callback to httplib that is called when the server enters the "running" state (similar in concept to jQuery's `$(document).ready(callback)`)
2. OR instead of a callback, write a function that returns `std::future<Server>`. This will allow threads to block when they attempt to read the running state of the Server through the future object, until the server has started (or failed to do so).
3. OR make the user write a spinlock on their side to constantly check `is_running()` and only call `stop()` when it returns `true`
4. OR change the `is_running()` function to return a reference to the atomic bool instead of a copied bool, so that users who happen to be using C++20 can use the `atomic::wait` function.

**Comment by yhirose** [comment]
@luismigb, hope @PixlRainbow's comment helps you.
